### DISCUSSION TITLE
Accessing contents

### DISCUSSION THREAD

**Discussion by jwalway** [cefsharp/CefSharp#4130] [Q&A]
Hello,

I tried signing up with the CeForum (https://magpcss.org/ceforum/index.php), but for some reason the email validations are not being sent to me.     I don't know if anyone can help me with that?

I wanted to ask (on there) if you can get the contents of a ChromiumWebBrowser control, by which I mean the contents of a webpage or webchat out of the browser by some some existing function or method in the control.  I have the control in Windows Forms .Net Framework 4.6.

Thanks!

**Comment by amaitland** [comment]
> I tried signing up with the CeForum (https://magpcss.org/ceforum/index.php), but for some reason the email validations are not being sent to me. I don't know if anyone can help me with that?

There are some details at https://bitbucket.org/chromiumembedded/cef/issues/2962/unable-to-register-in-your-forums#comment-57861660 

Past that I cannot help you.

>  ChromiumWebBrowser control, by which I mean the contents of a webpage or webchat out of the browser by some some existing function or method in the control.

What do you mean by contents exactly? If you mean the `HTML Source` then you can call [GetSourceAsync](https://cefsharp.github.io/api/101.0.x/html/M_CefSharp_WebBrowserExtensions_GetSourceAsync_1.htm). Make sure to add a `using CefSharp;` to your file.

`GetSourceAsync` is a snapshot in time, if your web page is dynamically generated the `HTML` returned will reflect the point in time at which it was called.

**Reply by jwalway** [reply]
Thanks for the reply!

> What do you mean by contents exactly? If you mean the HTML Source then you can call [GetSourceAsync](https://cefsharp.github.io/api/101.0.x/html/M_CefSharp_WebBrowserExtensions_GetSourceAsync_1.htm). Make sure to add a using CefSharp; to your file

Well, what I want to do is capture chat text continuously (say once per second).  The idea is to extract commands for a streaming game.   I can capture the text by clicking on the chat window, or next to it, then press "Ctrl+A" (to select all) and then "Ctrl+C" to copy the text to the clipboard.  After I can process the text in the clipboard.   I just want a more efficient way of getting the text.
.

**Reply by amaitland** [reply]
It sounds like what you want to do is query the `DOM`.

The two options are:

- Directly execute javascript and return the text you are after. See https://github.com/cefsharp/CefSharp/wiki/General-Usage#2-how-do-you-call-a-javascript-method-that-returns-a-result
- Use https://github.com/cefsharp/Puppeteer#cefsharp-puppeteer which provides a `C# API` 

>Well, what I want to do is capture chat text continuously (say once per second).

Polling generally isn't a great idea from a resource usage point of view, observing the changes using say a `MutationObserver` and calling `CefSharp.PostMessage` when the DOM changes is potentially better. https://github.com/cefsharp/CefSharp/wiki/Frequently-asked-questions#JSEvent

> then press "Ctrl+A" (to select all) and then "Ctrl+C" to copy the text to the clipboard. After I can process the text in the clipboard

For what it's worth you can programmatically call [SelectAll](https://cefsharp.github.io/api/101.0.x/html/M_CefSharp_WebBrowserExtensions_SelectAll_1.htm) and [Copy](https://cefsharp.github.io/api/101.0.x/html/M_CefSharp_WebBrowserExtensions_Copy_1.htm).

Generally speaking this is a little crude

**Reply by jwalway** [reply]
A lot of good information.  I'm going to have to study it to understand it better.

I have tried the SelectAll() and Copy(). The actual problem was to figure out how to put "focus" (usually done with a mouse click) to the right region of the window.    I understand it is crude.  I was just trying to find a way, since I don't have a full grasp of the technologies involved.  I'd love to find the best way, of course.

Thanks! I greatly appreciate your knowledge and your generosity with it! :-) 

...John

**Reply by amaitland** [reply]
The place to start is probably DevTools.

- https://developer.chrome.com/docs/devtools/
- https://github.com/cefsharp/CefSharp/wiki/Trouble-Shooting#javascript-debugging

**Reply by jwalway** [reply]
Hello amaitland,

I checked those links out.  They've been great guides. I was able to use the Inspector to search the DOM.  I found some good hooks for what I want to do.  That is, I used the **inspector** to execute JavaScript that gets the DOM elements I want.

  However, I ran into a roadblock which I haven't found the solution to.  I figure I can get the elements using the same code via **ExecuteScriptAync(**).  But it seems I must be doing something wrong at a basic level.  I experimented with JavaScript in my code.

I was able to execute this simple example:
`Browser.ExecuteScriptAsync("document.body.style.background = 'green';");`

That worked well.

Then I tried
`JavascriptResponse response = await Browser.EvaluateScriptAsync("1 + 1");`

And it returned 2 as expected.

  So, everything has looked good up to here.

However, when I tried to run basic **document** commands to return values, they returned nothing.  Example:
`JavascriptResponse response = await Browser.EvaluateScriptAsync(@"document.getElementsByTagName('a')");`

 even on benign websites with very little going on it says the response was a "**success"** but it returns "**null**" as a result. There are several href 'a' elements for it to extract on the pages, but it returns null.

I've tried several JavaScript functions with the same result, null.

In the Chrome Inspector everything works properly, as you'd expect.

 There aren't a lot of lines of code.   It's barebones.  So, I wonder what I'm doing wrong here.

**Reply by aimcosoftware** [reply]
This is a limitation of CEF. The function can only return simple types and classes, not DOM objects.
You would need to write the whole functionality in script and return a simple type.

**Reply by amaitland** [reply]
> Primitive data types int, double, date, bool and string are supported.
- Objects are supported to a degree and will be returned as IDictionary<string, object>, the use of the dynamic keyword is supported to make accessing the property values easier.
- You cannot directly return a DOM [Element](https://developer.mozilla.org/en-US/docs/Web/API/Element) (or anything with a cyclic reference), you need to create a new object that has just the information you need to return.
- Arrays containing the primitives listed above and objects are supported and will be returned as IList<object>.
- Array Like objects like [HTMLCollection](https://developer.mozilla.org/en-US/docs/Web/API/HTMLCollection) cannot be returned directly use [Array.from](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/from) to return and array

As per https://github.com/cefsharp/CefSharp/wiki/General-Usage#2-how-do-you-call-a-javascript-method-that-returns-a-result

There also some examples of returning a subset of properties of a htmlelement. 

Alternatively use CefSharp.Puppeteer

**Reply by jwalway** [reply]
Thanks aimcosoftware and amaitland.    I now have it working as I want.  _Except for one thing_.

I wanted to do this in my **Windows Forms** browser, but it wasn't recognizing javascript, so to speed things up I just did all of my experimenting in the **CefSharp.MinimalExample.Wpf.netcore** example.  That all worked.   So, now I want to get it working in the [SharpBrowser](https://github.com/sharpbrowser/SharpBrowser).    I can't make it recognize the javascript.    It will launch devtools, **ChromiumWebBrowser.ShowDevTools(),** but it gives this error when the dev tools window runs:

_"DevTools was disconnected from the page.
Once page is loaded, DevTools will automatically reconnect."_

  This is the error it keeps returning no matter if I reload the page or not.
 I wonder if the reason for the issues is that it's built on **.Net Framework 4.6**?

  I'd like to get this browser to recognize javascript if possible.

Thanks,
...John

**Reply by amaitland** [reply]
> So, now I want to get it working in the [SharpBrowser](https://github.com/sharpbrowser/SharpBrowser).

Make sure you are using the latest CefSharp version and all the required files are in the bin directory. 


https://github.com/cefsharp/CefSharp/wiki/Output-files-description-table-%28Redistribution%29

> I wonder if the reason for the issues is that it's built on .Net Framework 4.6?

No.

 .Net 452 or higher will work. The minimal example has a net 452 set of examples.

**Reply by jwalway** [reply]
Hello amaitland,

I made all of those changes and it still won't recognize javascript.  It was at version 89 and is now at 102.  In the process certain errors cropped up. The bin directory checks out that it has all of the files in the distribution.

The errors were that certain parameters are no longer valid for browser initialization:
One was this, which I commented out:
`CefSharpSettings.LegacyJavascriptBindingEnabled = true;`

I also had to commented out all of these:
```
config.FileAccessFromFileUrls = (!CrossDomainSecurity).ToCefState();
config.UniversalAccessFromFileUrls = (!CrossDomainSecurity).ToCefState();
config.WebSecurity = WebSecurity.ToCefState();
config.ApplicationCache = ApplicationCache.ToCefState();
```


It compiles and runs, but no javascript, it appears.  The same error with DevTools appears.

**Reply by amaitland** [reply]
Check the log file for errors see https://github.com/cefsharp/CefSharp/wiki/Trouble-Shooting#log-file

I don't have enough information to speculate what's going on.

**Reply by jwalway** [reply]
Here is the data the logfile produced right after I launched DevTools (**ChromiumWebBrowser.ShowDevTools()**):

> [0610/151723.074:VERBOSE1:network_delegate.cc(34)] NetworkDelegate::NotifyBeforeURLRequest: https://www.google.com/gen_204?atyp=csi&ei=pqajYu3fIvXDuvQPycSdyAg&s=webhp&st=44685&fid=5&t=fi&zx=1654892243070
> [0610/151723.077:VERBOSE1:tls_handshaker.cc(101)] TlsClient: Continuing handshake
> [0610/151723.080:VERBOSE1:tls_handshaker.cc(112)] TlsClient: SSL_do_handshake returned when entering early data. After retry, rv=-1, SSL_in_early_data=1
> [0610/151723.117:VERBOSE1:tls_handshaker.cc(101)] TlsClient: Continuing handshake
> [0610/151723.118:VERBOSE1:tls_handshaker.cc(101)] TlsClient: Continuing handshake
> [0610/151723.119:VERBOSE1:tls_client_handshaker.cc(482)] Client: handshake finished
> [0610/151724.386:VERBOSE1:network_delegate.cc(34)] NetworkDelegate::NotifyBeforeURLRequest: https://www.google.com/gen_204?atyp=i&ei=pqajYu3fIvXDuvQPycSdyAg&ct=slh&v=t1&im=M&pv=0.37036387928939285&me=17:1654892204513,V,0,0,0,0:18688,h,1,1,i:124,h,1,1,o:170,h,1,1,i:19030,h,1,1,o:216,h,1,1,i:247,V,0,0,932,591:78,G,1,1,79,123:200,h,1,1,o:1111,e,B&zx=1654892244378
> [0610/151724.710:VERBOSE1:media_stream_manager.cc(915)] RFAOSF::Core() [process_id=7, frame_id=1]
> [0610/151724.727:VERBOSE1:media_stream_manager.cc(915)] RFAOSF::Core() [process_id=7, frame_id=1]
> [0610/151724.786:VERBOSE1:script_context.cc(203)] Created context:
>   extension id:           (none)
>   frame:                  47C072C8
>   URL:                    
>   context_type:           WEB_PAGE
>   effective extension id: (none)
>   effective context type: WEB_PAGE
> [0610/151724.788:VERBOSE1:script_context.cc(203)] Created context:
>   extension id:           (none)
>   frame:                  00000000
>   URL:                    
>   context_type:           UNSPECIFIED
>   effective extension id: (none)
>   effective context type: UNSPECIFIED
> [0610/151724.789:VERBOSE1:dispatcher.cc(444)] Num tracked contexts: 1
> [0610/151724.959:INFO:cpu_info.cc(53)] Available number of cores: 8
> [0610/151726.041:VERBOSE1:CONSOLE(1)] "Main._createAppUI: 107.85888671875 ms", source: devtools://devtools/entrypoints/main/main.js (1)
> [0610/151726.542:VERBOSE1:CONSOLE(1)] "Main._showAppUI: 497.406005859375 ms", source: devtools://devtools/entrypoints/main/main.js (1)
> [0610/151726.637:VERBOSE1:CONSOLE(1)] "Main._initializeTarget: 42.3779296875 ms", source: devtools://devtools/entrypoints/main/main.js (1)
> [0610/151726.755:VERBOSE1:CONSOLE(1)] "Main._lateInitialization: 0.362060546875 ms", source: devtools://devtools/entrypoints/main/main.js (1)
>

**Reply by amaitland** [reply]
Try comment out all the handlers at https://github.com/sharpbrowser/SharpBrowser/blob/eeed5e8434a22cffa21c4962cfa36b11be38efc0/src/MainForm.cs#L379

**Reply by jwalway** [reply]
Hello amaitland,

  I commented them out and have been doing research on this problem, but still am having no luck.

When I attempt to run a script, this property **ChromiumWebBrowser.CanExecuteJavascriptInMainFrame** is continually _false_.


   Here are is a screenshot of the lines I commented out:
 

![jek0jak_commentOutSharpBrowserHandlers](https://user-images.githubusercontent.com/46982480/173198808-9a513bc4-d163-4224-bfc2-ff52b5e731b6.jpg)

**Reply by amaitland** [reply]
I'm not the developer of `SharpBrowser`, so not familiar with the code.

**Reply by jwalway** [reply]
Okay, thanks for the help, anyway.

Regards,
...John

**Reply by amaitland** [reply]
Is there a reason you want to use `SharpBrowser`? Some of it looks to be a bit of a mess

**Reply by jwalway** [reply]
I'm only choosing it because my friend can't get the other embedded browser (the one I had working with JavaScript above) to log onto TikTok.     He was able to get the SharpBrowser to log on.   So, I thought I'd see if I could get the JavaScript working in it with no success.    I am able to log onto TikTok with both browsers, myself.

So, I'm in a bind in that either way I go I can't make things work for him.  He's not super tech savvy, so it's kind of hard to get him to work through these things.   He has been trying, but I think it frustrates him.

Maybe there is an alternative CefSharp browser I don't know about?    I've been programming all of this in C Sharp.

**Reply by amaitland** [reply]
https://github.com/hardkoded/puppeteer-sharp

Run with headless = false and you should have a full chromium browser you can automate.

**Reply by jwalway** [reply]
Hey, thanks, amaitland!

I'll check that out!

**Reply by jwalway** [reply]
Well,   I've been playing around with Puppeteer and I attempted to use it to logon to TikTok, and the logon tells me "too many attempts," even though there is only one.   It's really adverse to automation software.

**Reply by amaitland** [reply]
It's very common for major websites to implement strategies to block embedded/automated browsers.
